test/e2e/registration_service_test.go

```diff
@@ -429,6 +432,138 @@ func (s *registrationServiceTestSuite) TestSignupOK() {
 	assert.Equal(s.T(), ExpectedConsoleURL(s.T(), s.memberAwait, memberCluster), mp["consoleURL"])
 }
 
+func (s *registrationServiceTestSuite) TestPhoneVerification() {
+	// Create a token and identity to sign up with
+	identity0 := authsupport.NewIdentity()
+	emailValue := uuid.NewV4().String() + "@some.domain"
+	emailClaim0 := authsupport.WithEmailClaim(emailValue)
+	token0, err := authsupport.GenerateSignedE2ETestToken(*identity0, emailClaim0)
+	require.NoError(s.T(), err)
+
+	// Call the signup endpoint
+	invokeEndpoint(s.T(), "POST", s.route + "/api/v1/signup", token0, "", http.StatusAccepted)
+
+	// Wait for the UserSignup to be created
+	userSignup, err := s.hostAwait.WaitForUserSignup(identity0.ID.String(), wait.UntilUserSignupHasConditions(PendingApproval()...))
+	require.NoError(s.T(), err)
+	emailAnnotation := userSignup.Annotations[v1alpha1.UserSignupUserEmailAnnotationKey]
+	assert.Equal(s.T(), emailValue, emailAnnotation)
+
+	// Call get signup endpoint with a valid token and make sure verificationRequired is true
+	mp, mpStatus := parseResponse(s.T(), invokeEndpoint(s.T(), "GET", s.route + "/api/v1/signup", token0, "", http.StatusOK))
+	assert.Equal(s.T(), "", mp["compliantUsername"])
+	assert.Equal(s.T(), identity0.Username, mp["username"])
+	require.IsType(s.T(), false, mpStatus["ready"])
+	assert.False(s.T(), mpStatus["ready"].(bool))
+	assert.Equal(s.T(), "PendingApproval", mpStatus["reason"])
+	require.True(s.T(), mpStatus["verificationRequired"].(bool))
+
+	// Confirm the status of the UserSignup is correct
+	_, err = s.hostAwait.WaitForUserSignup(identity0.ID.String(), wait.UntilUserSignupHasConditions(NotApprovedAndVerificationRequired()...))
+
+	// Confirm that a MUR hasn't been created
+	obj := &v1alpha1.MasterUserRecord{}
+	err = s.hostAwait.Client.Get(context.TODO(), types.NamespacedName{Namespace: s.hostAwait.Namespace, Name: identity0.Username}, obj)
+	require.Error(s.T(), err)
+
+	// Initiate the verification process
+	invokeEndpoint(s.T(), "PUT", s.route + "/api/v1/signup/verification", token0,
+		`{ "country_code":"+61", "phone_number":"408999999" }`, http.StatusNoContent)
+
+	// Retrieve the updated UserSignup
+	userSignup, err = s.hostAwait.WaitForUserSignup(identity0.ID.String())
+	require.NoError(s.T(), err)
```

Comment:
If I understand it correctly, we send `PUT` request to the registration service with the country code and phone number and the reg-service then updates the UserSignup. I know that it may look like a detail, but it could happen that we get the UserSignup in the test earlier than registration service updates it - I was just fixing such flaky tests https://github.com/codeready-toolchain/toolchain-e2e/pull/172 so I'd like to avoid it for the future. 
WDYT? Could you create a `UserSignupWaitCriterion` for that?

Reply:
I'm not quite sure what you mean here.  The init verification (i.e. the PUT request) is synchronous, it will not return a response until after the UserSignup has been updated.  The `invokeEndpoint()` function is also synchronous, it will not return until it has fully read the body of the response.  How is it possible that we could get the `UserSignup` in the test earlier than the registration service updates it?

Reply:
ah, right - I totally forgot that it is synchronous (I'm still living in the asynchronous world :grimacing:) Thanks for explanation :+1: